netket/utils/group/group.py

```diff
@@ -0,0 +1,266 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Ignore false-positives for redefined `product` functions:
+# pylint: disable=function-redefined
+
+import numpy as np
+from dataclasses import dataclass
+
+from .semigroup import SemiGroup, Element, Identity
+from netket.utils import struct, HashableArray
+from netket.utils.float import comparable, prune_zeros
+from netket.utils.types import Array, DType, Shape
+from typing import Tuple, List
+
+
+@struct.dataclass
+class Group(SemiGroup):
+    """
+    Collection of Elements expected to satisfy group axioms.
+    Unlike SemiGroup, product tables, conjugacy classes, etc. can be calculated.
+
+    Group elements can be implemented in any way, as long as a subclass of Group
+    is able to implement their action. Subclasses must implement a :code:`_canonical()`
+    method that returns an array of integers for each acceptable Element such that
+    two Elements are considered equal iff the corresponding matrices are equal.
+
+    """
+
+    def __hash__(self):
+        return super().__hash__()
+
+    def __matmul__(self, other) -> "Group":
+        if not isinstance(other, Group):
+            raise ValueError("`Group`s can only be multiplied with other `Group`s")
+        return Group(super().__matmul__(other).elems)
```

Comment:
let's make this with dispatch.
I also think it's legit to multiply a group by a semigroup.
you get a semigroup as a result.

Reply:
How can you write a dispatch so that it won't conflate group-element multiplication with group multiplication?

Reply:
The first step is moving group matmul to `product`.
I would do this:
 - remove `Group.__matmul__`
 - change `SemiGroup.__matmul__(self, other)` to call `product(self, other)`
 - implement the following dispatch rules:

```python
@dispatch
def product(A: SemiGroup, B: SemiGroup):
    return SemiGroup(elems=[a @ b for a, b in itertools.product(A.elems, B.elems)])

@dispatch
def product(A: Group, B: Group):
    return Group(elems=[a @ b for a, b in itertools.product(A.elems, B.elems)])

# And those are not strictly needed but might be usefull
@dispatch
def product(A: SemiGroup, B: Group):
    return SemiGroup(elems=[a @ b for a, b in itertools.product(A.elems, B.elems)])

@dispatch
def product(A: Group, B: SemiGroup):
    return SemiGroup(elems=[a @ b for a, b in itertools.product(A.elems, B.elems)])
```

Reply:
I guess the last two follow because `Group` subclasses `SemiGroup`?

Reply:
Do you want this to be the same dispatched `product`as that between `Element`s? Should we extend that to be `Element @ Group` as well?

Reply:
> I guess the last two follow because Group subclasses SemiGroup?

You're right indeed. I had not noticed.

> productas that between Elements? Should we extend that to be Element @ Group as

Yes that's my idea.  Though I'd leave it for later. If that uses dispatch than adding more rules is very fast and easy
We'd have to check if Element is in group,  then return the group itself else do all the multiplications...

Reply:
> Yes that's my idea. Though I'd leave it for later. If that uses dispatch than adding more rules is very fast and easy

OK, it's not something we desperately need anyway

> We'd have to check if Element is in group, then return the group itself else do all the multiplications...

For the record that's not a great idea, even if you only reorder the group, that would be my expected behaviour from a Cartesian product operator